### PR TITLE
cool#9621: Have WSD Keep Download-Socket open for file and favicon.ico requests (ClientRequestDispatcher); ..

### PR DESCRIPTION
**Pull request by sgothel**
### Summary

WSD's ClientRequestDispatcher shall keep the socket open on successful document and favicon download requests.

HttpHelper::sendFile(..) is refactored from HttpHelper::sendFileAndShutdown(..) w/o the socket closing (shutdown) and used in WSD's ClientRequestDispatcher accordingly.

WebSocketHandler: Added public isConnected() StreamSocket queries, allowing our unit tests to validate the connected state.

#### Keep-Alive Utilization
FileServer.cpp's FileServerRequestHandler::handleRequest(..) also keeps the socket alive (ENABLE_DEBUG and env 'COOL_SERVE_FROM_FS') with this patch.

Check whether this is desired, as not covered by a unit test!

+++

Analyze whether the following maintained HttpHelper::sendFileAndShutdown(..) use-cases shall be converted to keep-alive as well:
- WSD's ClientSession::handleKitToClientMessage()

+++

Unit tests w/ WopiTestServer still use HttpHelper::sendFileAndShutdown(..):
- test/UnitWOPIFileUrl.cpp:145
- test/UnitWOPITemplate.cpp:81

Currently the whole WopiTestServer code closes the sockets even after serving a request successfully.
We might want to change this behavior towards keep-alive as well? This would also allow us to employ 'local' performance unit tests with such a WopiTestServer.

#### Poco HTTP Message Parsing
Replace Poco http request message parsing
starting with Ash's net/HttpRequest

Change-Id: I83c09e0a8101931efece42e42e5a8e3abd7e696f

* Resolves: #9621
* Target version: master 

### Checklist

- [x] I have run `make prettier-write` and formatted the code.
- [x] All commits have Change-Id
- [x] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: test/UnitSession.cpp
Comment (mmeeks) [inline test/UnitSession.cpp:253]: We can run the socket logic on the current thread if we set:
    bool runOnClientThread()
on the SocketPoll - which can make the tests simpler to follow.
Use 'SocketPoll::poll()' in a loop to wait for your condition then and avoid threads / races =)

File: test/UnitSession.cpp
Comment (mmeeks) [inline test/UnitSession.cpp:279]: We should assert the non-presence of the "Connection: close" header I think.

File: test/UnitSession.cpp
Comment (mmeeks) [inline test/UnitSession.cpp:258]: Surely we should use Ash's new HTTP code here and elsewhere.

Also we don't need to load a document in order to do these HTTP checks. I'd suggest we don't and instead focus on the various kinds of special files we serve: capabilities, prometheus metrics etc. etc.

**Review by mmeeks — COMMENTED** [review]
Looks good generally, would be great to ensure that we are not keeping connections open and sending "Connection: close" down them - and of course, we need to have a follow-on that audits all the sites where we sendFileAndShutdown which I expect are many and varied =)

File: test/UnitSession.cpp
Comment (sgothel) [reply test/UnitSession.cpp:279]: `Connection: close` in the http header seems to be for client request only, signaling the server that no 'keep alive' is desired. 
While such a server response doesn't seem to be viable or possible here, we might want to use it in the WSD request handling to keep the connection open (as per this CR/PR) IFF not set. 
Hence we would still close the socket in the WSD if client optionally requests `Connection: close`, I earmark this and check with the WSD ClientRequestDispatcher code.

File: test/UnitSession.cpp
Comment (sgothel) [reply test/UnitSession.cpp:258]: I will add more test cases and expand its use case, i.e. covering the noted `FileServerRequestHandler::handleRequest(..)` and learning about the different served files. Thx!

Probably revisiting the unit test `UnitWOPIFileUrl` if suitable, as I tossed this one out after my initial attempt.
IMHO I will use the WOPI file request as used in `UnitWOPIFileUrl`, but repackage it in this linear UnitSession logic w/o the provided  WopiTestServer code.

Comment (sgothel) [reply test/UnitSession.cpp:258]: .. also will add the new HTTP code usage on top of the revised patch as intended earlier, to have the commits more 'on the point'.

File: test/UnitSession.cpp
Comment (sgothel) [reply test/UnitSession.cpp:253]: looking into it.. thx

File: test/UnitSession.cpp
Comment (sgothel) [reply test/UnitSession.cpp:279]: note: this unit tests checks whether the 2 sockets are still open in subsequent requests.

File: test/UnitSession.cpp
Comment (sgothel) [reply test/UnitSession.cpp:258]: yeah, while working on the changeset .. I also realized its better to put the new HTTP code in and enhance the http::Header with get/set methods for `Connection`. All in progress.

File: net/HttpRequest.hpp
Comment (mmeeks) [inline net/HttpRequest.hpp:467]: Content-Length    = "Content-Length" ":" 1*DIGIT

Surely - we don't expect the content_length token to be anything but a number ? but I didn't read the get() impl.

File: test/UnitSession.cpp
Comment (mmeeks) [inline test/UnitSession.cpp:198]: Better =)

**Review by mmeeks — COMMENTED** [review]
Looks plausible; but I'm going to be out from tomorrow for some time - so Ash or Caolan will need to read/review updates.

File: net/HttpRequest.hpp
Comment (sgothel) [reply net/HttpRequest.hpp]: ;-) .. reading your comment after my fix. Yeah: CONNECTION it is now. *slap* myself.
https://github.com/CollaboraOnline/online/commit/9998e90717a8a4083cdec3b82c9cc6d95e57634d

**Comment by sgothel** [comment]
Last 2 commits reflect input given from @Ashod . 

TODO: The actual FileServer handling `FileServerRequestHandler::handleRequest(..)` 
.. coming up. Should be the last piece for this pull request.

**Comment by sgothel** [comment]
rebased to master 42e9dac882d0558307d49c8d739726ab5691ea3a

**Comment by sgothel** [comment]
OK, last commits done, rebased and forced pushed .. ready. @caolanm @Ashod @mmeeks

File: net/HttpRequest.hpp
Comment (Ashod) [inline net/HttpRequest.hpp:472]: Ditto as above. No need for all these `else if` after returning. Just have these as `if/return` cases, without all the `else`s. More readable as the structure of the code reflects the intent and functionality.

File: net/HttpRequest.hpp
Comment (Ashod) [inline net/HttpRequest.hpp:480]: Nit: blank line between clauses and functions.

File: net/HttpRequest.hpp
Comment (Ashod) [inline net/HttpRequest.hpp:483]: Nit: More readable to break these on multiple lines. The reader's time and attention are more valuable resources than a line or two on the screen (let alone in the file). (Don't force me to parse 3 statements on one line, please.)

File: net/HttpRequest.hpp
Comment (Ashod) [inline net/HttpRequest.hpp:1479]: Somehow the space between the `if` and the paren went away. `if`, `for`, `while`, etc. are not functions. It's customary to leave a space to differentiate between them and function calls.

(Again, making it easier for maintainers to read and parse code quickly and will lower chance of confusion/misreading is a good thing.)

**Review by Ashod — COMMENTED** [review]
Thanks, @sgothel. Good stuff. I left a few more comments, but most of my feedback was communicated offline, so not much new here.

One thing I wonder is how long the browser keeps this connection open. Could it be indefinite (as long as the tab is open) ? I think we should check with different browsers, so we can timeout and close them if they end up lingering for too long. We don't serve http files after loading the document, but the browser doesn't know that.

Of course CI also needs to be green, but I didn't check why there are failures. Let me know if need help looking into the reasons.

File: wsd/ClientRequestDispatcher.cpp
Comment (caolanm) [inline wsd/ClientRequestDispatcher.cpp:767]: cypress tests pass/fail with this shutdown present/omitted

**Review by caolanm — CHANGES_REQUESTED** [review]
try the links from the "make run" prompt and they have the same issue I believe

File: wsd/ClientRequestDispatcher.cpp
Comment (caolanm) [reply wsd/ClientRequestDispatcher.cpp:767]: using:

if (request.getMethod() == Poco::Net::HTTPRequest::HTTP_POST)
    socket->shutdown();

here I think worksaround the problem, which I think is that ClientRequestDispatcher::handleIncomingMessage is using socket->eraseFirstInputBytes(map); to remove consumed data, but in the post we have body content that isn't consumed by that and gets left behind

File: net/Socket.cpp
Comment (Ashod) [inline net/Socket.cpp:1309]: Why convert micros into fractional millis? Seems having duration values as microsecond integers is more straightforward than this conversion.

File: net/Socket.cpp
Comment (Ashod) [inline net/Socket.cpp:1320]: If `delayMs` is in a standard unit (i.e. `chrono::millseconds`, `chrono::microseconds`, etc.), you don't need the `.count()` nor the `<< "ms"`, instead, just stream `delayMs` and you get the units too! (Yet another reason to avoid floating-point duration).

File: net/Socket.cpp
Comment (Ashod) [inline net/Socket.cpp:1434]: Are these exceptions coming from the form parser? I'm rather surprised, if so.

File: net/Socket.cpp
Comment (Ashod) [inline net/Socket.cpp:1310]: I expect we'd want a separate value for the `delayMax`, other than `DefaultPollTimeoutMicroS`, as I don't see how a socket-specific timeout would be related to our poll-timeout, which can be arbitrarily small or long (indeed, it can even be infinite!).

File: wsd/ClientRequestDispatcher.cpp
Comment (Ashod) [inline wsd/ClientRequestDispatcher.cpp:767]: Would be good to have a comment explaining what's going on here. This is completely non-intuitive and we spent quite a bit of time to figure this workaround out. I know I'll likely forget what this is doing in a few short weeks.

**Review by Ashod — APPROVED** [review]
Thanks @sgothel. Left some comments and suggestions, but they can be addressed in a follow up. I think this is good to go in, if @caolanm is happy.

**Review by caolanm — APPROVED** [review]

**Comment by caolanm** [comment]
yeah, lets do follow ups separately and get this in for maximum testing exposure

**Pull request merged**
